hasher-matcher-actioner/hmalib/lambdas/actions/action_performer.py

```diff
@@ -1,33 +1,96 @@
-    that aggregates labels and 'decides' which action would be taken.
-
-    For now, this method will,
-    - it will construct the MatchMessage object and identify the actions it
-      needs to invoke.
-    - for now, rather than fanning out to individual specific lambdas, it will
-      call all specific functions serially!
+    This lambda is called when one or more matches are found. If a single hash matches
+    multiple datasets, this will be called only once.
+
+    Action labels are generated for each match message, then an action is performed
+    corresponding to each action label.
     """
-    for sqs_record in event["Records"]:
+    for sqs_record in event["Records"]: #confused about this... are we popping off more than one thing at a time from the "matches" queue?
         sns_notification = json.loads(sqs_record["body"])
         match_message: MatchMessage = MatchMessage.from_sns_message(
             sns_notification["Message"]
         )
+        action_labels = get_action_labels(match_message)
+        for action_label in action_labels:
+            perform_action(match_message, action_label)
+
+        # the more I write code here about reacting to ThreatExchange, the more I wish it was somewhere else entirely. Implementing
+        # here forever ties up action-related code with ThreatExchange-specific code / functionality. I prefer putting
+        # ThreatExchange-related specifics for reacting into another lambda that also subscribes to the matcher's notifications.
+        if threat_exchange_reacting_is_enabled():
+            threat_exchange_reaction_labels = get_threat_exchange_reaction_labels(action_labels)
+            if threat_exchange_reaction_labels is t.List[Label]:
+                for threat_exchange_reaction_label in threat_exchange_reaction_labels:
+                    react_to_threat_exchange(match_message, threat_exchange_reaction_label)
+
+
+def get_action_labels(match_message: MatchMessage) -> t.List[ActionLabel]:
+    """
+    TODO finish implementation
+    Returns an ActionLabel for each ActionRule that applies to a MatchMessage. 
+    """
+    action_rules = get_action_rules()
+    action_labels = [] # use Set here
+    for action_rule in action_rules:
+        if action_rule.applies(match_message) and not action_labels.contains(action_rule.action_label):
+            action_labels.append(action_rule.action_label)
+    action_labels = remove_superseded_actions(action_labels) # maybe not needed for v0
+    return action_labels
+
+
+def get_action_rules() -> t.List[ActionRule]:
+    """
+    TODO implement
+    Returns a collection of ActionRule objects. An ActionRule will have the following attributes:
+    MustHaveLabels, MustNotHaveLabels, ActionLabel
+    """
+    return [ActionRule(ActionLabel("Action", "EnqueueForReview"), [Label("Collaboration", "12345")], [])]
+
+
+def get_actions() -> t.List[Action]:
+    """
+    TODO implement
```

Comment:
And this will be fetched from the database, right? That's where the rules live?

Reply:
@schatten I'm wondering about that. I was going to wait to hear from @Dcallies before making any decision... Seems like if we've got a single place for all our configurations, rules, actions, webhook details, pretty much anything might land in the common config data store that might emerge.